linera-execution/src/wasm/common.rs

```diff
@@ -34,165 +35,60 @@ pub trait ApplicationRuntimeContext: Sized {
-    type HandleSessionCall: GuestFutureInterface<
-            Self,
-            Output = (SessionCallResult, Vec<u8>),
-            Parameters = (CalleeContext, Vec<u8>, Vec<u8>, Vec<SessionId>),
-        > + Send
-        + Unpin;
-
-    /// The WIT type eqivalent for [`Poll<Result<RawExecutionResult<Vec<u8>>, String>>`].
-    type PollExecutionResult;
-
-    /// The WIT type eqivalent for [`Poll<Result<ApplicationCallResult, String>>`].
-    type PollApplicationCallResult;
-
-    /// The WIT type eqivalent for [`Poll<Result<SessionCallResult, String>>`].
-    type PollSessionCallResult;
-
-    /// Creates a new future for the user application to initialize itself on the owner chain.
-    fn initialize_new(
+    /// Initializes the user application on the owner chain.
+    fn initialize(
         &self,
         store: &mut Self::Store,
         context: OperationContext,
         argument: Vec<u8>,
-    ) -> Result<Self::Initialize, Self::Error>;
-
-    /// Polls a user contract future that's initializing the application.
-    fn initialize_poll(
-        &self,
-        store: &mut Self::Store,
-        future: &Self::Initialize,
-    ) -> Result<Self::PollExecutionResult, Self::Error>;
+    ) -> Result<Result<RawExecutionResult<Vec<u8>>, String>, Self::Error>;
 
-    /// Creates a new future for the user application to execute an operation.
-    fn execute_operation_new(
+    /// Executes an application's operation.
+    fn execute_operation(
         &self,
         store: &mut Self::Store,
         context: OperationContext,
         operation: Vec<u8>,
-    ) -> Result<Self::ExecuteOperation, Self::Error>;
+    ) -> Result<Result<RawExecutionResult<Vec<u8>>, String>, Self::Error>;
 
-    /// Polls a user contract future that's executing an operation.
-    fn execute_operation_poll(
-        &self,
-        store: &mut Self::Store,
-        future: &Self::ExecuteOperation,
-    ) -> Result<Self::PollExecutionResult, Self::Error>;
-
-    /// Creates a new future for the user contract to execute a message.
-    fn execute_message_new(
+    /// Executes a cross-chain message.
+    fn execute_message(
         &self,
         store: &mut Self::Store,
         context: MessageContext,
         message: Vec<u8>,
-    ) -> Result<Self::ExecuteMessage, Self::Error>;
-
-    /// Polls a user contract future that's executing a message.
-    fn execute_message_poll(
-        &self,
-        store: &mut Self::Store,
-        future: &Self::ExecuteMessage,
-    ) -> Result<Self::PollExecutionResult, Self::Error>;
+    ) -> Result<Result<RawExecutionResult<Vec<u8>>, String>, Self::Error>;
 
-    /// Creates a new future for the user contract to handle a call from another contract.
-    fn handle_application_call_new(
+    /// Handles a call from another contract.
+    fn handle_application_call(
         &self,
         store: &mut Self::Store,
         context: CalleeContext,
         argument: Vec<u8>,
```

Comment:
Would a `&[u8]` work now, here and in the other places?

Reply:
Same for the `*Context` arguments

Reply:
No, unfortunately. The parameters are still sent to a different thread (inside `run_wasm_guest`), so they need to be `Send + 'static`.